#4562: add a query that includes BE filering of subject metadata and …

cpcop opened the pull request:
…session dates and count

cpcop committed:
#4562: add a query that includes BE filering of subject metadata and session dates and count

cpcop commented:
@lucazonta-moveshelf , to make the BE filters available to our customers, I have implemented a query that uses the new BE filters. Could you please review?. I have also added a query to retrieve all data from a particular subject (useful for the scripts to perform queries for research) (FYI @martinSchepers ). Since these changes are quite significant, I have bumped the version to from 1.5.2 to 1.6.0 (please let me know if you agree)

martinSchepers commented:
Did a quick test with the queries, and they worked nicely for me!

cpcop committed:
#4562: retrieve clips at session level and not at report level

cpcop commented:
discussed with @martinSchepers to retrieve clips at session level instead of at report level. See https://github.com/moveshelf/moveshelf-python-api/pull/13/commits/35ae0b4e2c56dcb4be954832974211bb73ea1455

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            return
+        if not isinstance(base_filter, dict):
+            raise ValueError("BaseFilterInput must be a dict or None.")
+        # Single filter: must have 'key', 'operator', 'value'
```
add this check

```python
    # Check for invalid mixed structure
    has_filter_fields = any(filter_dict.get(field) is not None for field in ["key", "operator", "value"])
    has_logic_fields = any(filter_dict.get(field) is not None for field in ["logic", "filters"])

    if has_filter_fields and has_logic_fields:
        raise ValueError(
            "Cannot have MetadataFilter fields (key/operator/value) and FilterLogic fields (logic/filters) together"
        )
```

cpcop replied:
done. Please note that I changed `any `to `all`, as all fields need to be present for the filter to be functional

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+        if not isinstance(base_filter, dict):
+            raise ValueError("BaseFilterInput must be a dict or None.")
+        # Single filter: must have 'key', 'operator', 'value'
+        if "key" in base_filter:
```
```suggestion
        if has_filter_fields:
```

cpcop replied:
done

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if not isinstance(base_filter["value"], str):
+                raise ValueError("'value' must be a string.")
+        # Logic group: must have 'logic' and 'filters'
+        elif "logic" in base_filter:
```
```suggestion
        elif has_logic_fields:
```

cpcop replied:
done

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+        if "key" in base_filter:
+            allowed_operators = ["EQ"]  # Extend if backend supports more
+            if "operator" not in base_filter:
+                raise ValueError("Single filter must have 'operator'.")
```
if you check `has_filter_fields` you can remove this

cpcop replied:
done

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if base_filter["operator"] not in allowed_operators:
+                raise ValueError(f"Invalid operator: {base_filter['operator']}. List of allowed operators: {allowed_operators}")
+            if "value" not in base_filter:
+                raise ValueError("Single filter must have 'value'.")
```
if you check `has_filter_fields` you can remove this

cpcop replied:
done

lucazonta-moveshelf commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,225 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if base_filter["logic"] not in allowed_logics:
+                raise ValueError(f"Invalid logic: {base_filter['logic']}. List of allowed logics: {allowed_logics}")
+            if "filters" not in base_filter:
+                raise ValueError("Logic group must have 'filters'.")
```
if you check `has_logic_fields` you can remove this

cpcop replied:
done

cpcop committed:
#4562: addressed review suggestions

cpcop commented:
@lucazonta-moveshelf I have addressed your comments. Please also see my note https://github.com/moveshelf/moveshelf-python-api/pull/13#discussion_r2533443510. Ready for review

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+
+        if has_filter_fields and has_logic_fields:
+            raise ValueError(
+                "Cannot have MetadataFilter fields (key/operator/value) and FilterLogic fields (logic/filters) together"
```
```suggestion
                "Cannot have metadata filter fields (key/operator/value) and filter logic fields (logic/filters) together"
```

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            for f in filter_dict["filters"]:
+                self._validate_filter_dict_input(f)
+        else:
+            raise ValueError("Input dictionary must have either MetadataFilter fields (key/operator/value) or FilterLogic fields (logic/filters).")
```
```suggestion
            raise ValueError("Input dictionary must have either metadata filter fields (key/operator/value) or filter fields (logic/filters).")
```

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+        if filter_dict is None:
+            return
+        if not isinstance(filter_dict, dict):
+            raise ValueError("BaseFilterInput must be a dict or None.")
```
```suggestion
            raise ValueError("metadataFilter dict must be a dict or None.")
```

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if session_filters is None:
+                return
+            if not isinstance(session_filters, dict):
+                raise ValueError("session_filters must be a dict or None.")
```
```suggestion
                raise ValueError("sessionFilter must be a dict or None.")
```

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if "sessionDates" in session_filters:
+                sd = session_filters["sessionDates"]
+                if not isinstance(sd, dict):
+                    raise ValueError("session_filters['sessionDates'] must be a dict.")
```
```suggestion
                    raise ValueError("sessionFilter.sessionDates must be a dict.")
```

zontaluca commented on src/moveshelf_api/api.py:
```diff
@@ -1236,6 +1236,226 @@ def getProjectSubjectByEhrId(self, ehr_id, project_id):
+            if "numSessions" in session_filters:
+                ns = session_filters["numSessions"]
+                if not isinstance(ns, dict):
+                    raise ValueError("session_filters['numSessions'] must be a dict.")
```
```suggestion
                    raise ValueError("sessionFilter.numSessions must be a dict.")
```

zontaluca reviewed (commented):
Looks good, but i changed few error messages

cpcop committed:
Apply suggestions from code review

Co-authored-by: Luca Zonta <zonta.luca92@gmail.com>

zontaluca reviewed (approved):
LGTM!
